README.md
# PC-JQ模板

![version](https://img.shields.io/github/package-json/v/hejingmiao/pc-jq-template.svg)
![commit](https://img.shields.io/github/last-commit/hejingmiao/pc-jq-template.svg)
![new feature](https://img.shields.io/badge/author-hejingmiao-orange)

<!-- ![code-size](https://img.shields.io/github/languages/code-size/hejingmiao/pc-jq-template.svg) -->
<!-- ![lang](https://img.shields.io/github/languages/top/hejingmiao/pc-jq-template.svg) -->

### 开发
```html
<!-- 添加cms_id，「0001」频道ID依据创建项目时的频道名,「xxxx」为发布器对应频道新建专题名，需要发布到发布器的项目，该项必填 -->
<meta name="cms_id" content="0001xxxx" />
<!-- 公共导航头 -->
<!--#include virtual="/special/ntes_common_model/nte_commonnav2019.html" -->
<!-- 频道导航头 -->
<!--#include virtual="/special/sp/post_1603_header.html" -->
```
### 安装

```bash
$ npx ne-build pc
# node版本大于5.2
# 参数可在命令行输入，也可以输入命令后按交互提示输入

# 参数⬇️
-n or --name
# 项目名称，必须
# 用于替换模板项目中package.json的{name}

-c or --channel
# 频道名称，必须
# 用于替换模板项目内上传路径中的频道路径

-t or --template
# 项目模版，必须

-d or --desc
# 项目描述，可选
# 用于替换模板项目中package.json的{description}

--username
# 用户名

--password
# 密码
```

### 用户配置
```js
// gulpfile.js
const account = require('/Users/xxx/code/openID') // 配置路径
```

### 命令
```bash
# 安装依赖
$ npm i

# 开启本地服务
$ npm run dev

# 初始化openid（只需执行一次）
$ npm run initopenid

# 打包（上传前需先打包）
$ npm run build

# 上传静态资源
$ npm run upload

# 发布cms
$ npm run publish

# 打包+上传+发布+git提交
npm run submit
```

### 目录结构

```
├── CHANGELOG.md
├── README.md
├── babel.config.js             # babel配置文件
├── config/
├── openID.json                 # 上传账号密码文件
├── package-lock.json
├── package.json
├── postcss.config.js           # postcss配置文件
├── script                      # webpck配置、上传发布脚本、初始化openID脚本
|  ├── init-openid.js
|  ├── publish.js
|  ├── upload-tools.js
|  ├── upload.js
|  ├── webpack.base.config.js
|  ├── webpack.dev.config.js
|  └── webpack.prod.config.js
├ src/
│  ├ assets/               # 图片等资源文件夹
│  ├ css/
│  │  ├ common.less        # 公用reset css
│  │  └ index.less         # 样式文件
│  ├ js/
│  │  ├ utils/             # 工具js文件夹
│  │  │  ├ helper.js       # eruda、统计js
│  │  │  ├ track.js        # 统计方法
│  │  │  └ utils.js        # 工具函数
│  │  ├ common.js          # 通用js
│  │  └ index.js           # js文件
├ static/                  # 静态资源文件夹
└── share-icon.png
```
### 简介

* 项目使用webpack打包，webpack-dev-server开启本地服务
* html默认使用.html，可按照[html-webpack-plugin文档](https://github.com/jantimon/html-webpack-plugin)根据需求自行配置模板文件。
* css默认使用[less](http://lesscss.org/)以及[postcss](https://postcss.org/)
* js使用[babel7.5](https://babeljs.io/)进行编译，按照package.json中的browserslist自动增加polyfill
* eslint使用默认标准，[规则](https://eslint.org/docs/rules/)
* 统计ID位于`package.json`的`projectId`字段
* 已内置`jQuery`，版本`v3.4.1`index.html中引入js，webpack.externals进行排除）
* npm run dev开启本地服务后，可以使用localhost:8080或{ip}:8080或dev.f2e.163.com:8080（需配置hosts）调试
* 打包时静态资源默认添加hash，上传后进行缓存，无修改不再重复上传
* 如需引用频道include，需要再webpack.dev.config.js中59行修改频道域名

CHANGELOG.md
<!--
 * @Author: wanghaiyan
 * @Date: 2022-02-08 10:15:16
 * @LastEditTime: 2022-02-11 14:23:22
 * @LastEditors: wanghaiyan
-->
# Changelog
## [Unreleased]

## [1.0.0] - 2020-02-21
### Added
- 基础版本

## [2.0.0] - 2022-02-11
### 修改上传和发布的方式


script/init-openid.js
/*
 * @Author: wanghaiyan
 * @Date: 2022-01-14 15:06:25
 * @LastEditTime: 2022-02-08 11:34:51
 * @LastEditors: wanghaiyan
 */
const path = require('path');
const fs = require('fs');
const inquirer = require('inquirer');

(async function run() {
  const target = path.join(__dirname, '../openID.json')
  if (!fs.existsSync(target)) {
    console.log('不存在' + target + '配置文件，准备生成..')
    const username = (await inquirer.prompt({name: 'user',message:'输入邮箱前缀'})).user
    const password = (await inquirer.prompt({ name: 'pwd', message: '输入openid密码', type:'password'})).pwd
    const config = {
      username,
      password,
      "cms": {
        "pc": {
          "url": "https://newcms.ws.netease.com/openapi/handleModelData",
          "params": "target=model&userid=f2e&password=abc%231234&extname=.html"
        },
        "_3g": {
          "url": "https://3gcms.ws.netease.com/cms/model/updateContent.do",
          "params": "target=model&userid=xqwei&password=D0erwxpw!!"
        }
      },
      otppwd: "将军令(外网环境下)"
    }
    fs.writeFileSync(target, JSON.stringify(config, null, 2));
    console.log('生成openID.json在' + target)
  }

}())

script/webpack.base.config.js
const path = require('path')
const resolve = path.resolve

const MiniCssExtractPlugin = require('mini-css-extract-plugin')

const dev = process.env.NODE_ENV !== 'production'

module.exports = {
  context: resolve(__dirname, '../'),
  entry: [
    path.join(__dirname, '../src/js/index.js')
  ],
  module: {
    rules: [
      {
        test: /\.(le|c)ss$/,
        use: [
        dev ? 'style-loader' : {
          loader: MiniCssExtractPlugin.loader,
          options: {
            publicPath: '../../'
          }
        },
        'css-loader',
        'postcss-loader',
        'less-loader',
        ]
      },
      {
        enforce: 'pre',
        test: /\.js$/,
        exclude: /node_modules/,
        include: [path.resolve(__dirname, 'src')],
        loader: 'eslint-loader',
        options: {
          formatter: require('eslint-friendly-formatter')
        }
      },
      {
        test: /\.js$/,
        exclude: /node_modules|\.min\.js|bower_components/,
        use: {
          loader: 'babel-loader'
        }
      },
      {
        test: /\.(png|jpe?g)(\?.*)?$/,
        use: [
          {
            loader: 'url-loader',
            options: {
              name: 'static/images/[name].[hash:7].[ext]',
              limit: 1024,
              fallback: 'file-loader'
            }
          },
          {
            loader: 'tinify-loader',
            options: {
              apikey: 'iA4WgA6dpM0nbSKsByJDA0MuLyodD2_j',
              cache: path.resolve(__dirname, 'node_modules/tinify-loader')
            }
          }
        ]
      },
      {
        test: /\.(gif|svg)(\?.*)?$/,
        use: [{
          loader: 'url-loader',
          options: {
            name: 'static/images/[name].[hash:7].[ext]',
            limit: 1024,
            fallback: 'file-loader'
          }
        }]
      },
      {
        test: /\.(mp3|mp4)(\?.*)?$/,
        loader: 'url-loader',
        options: {
          name:'static/[name].[ext]',
          limit:10
        }
      },
      {
        test: /\.(woff2?|eot|ttf|otf)(\?.*)?$/,
        loader: 'url-loader',
        options: {
          limit: 10000,
          name: './static/fonts/[name].[ext]'
        }
      }
    ]
  },
  externals: [
    /jquery|\$/
  ],
  resolve: {
    extensions: ['.js', '.json'],
    // 配置项目文件别名
    alias: {
      '@': resolve('src'),
      'utils': resolve('src/js/utils'),
      'js-bridge': '@mf2e/js-bridge',
      'newsapp-share': '@newsapp-activity/newsapp-share'
    }
  // },
  // resolveLoader: {
  //   modules: [path.join(__dirname, '../ne-velocity-loader'), 'node_modules']
  }
}

script/webpack.prod.config.js
const pkg = require('../package.json')
const baseWebpackConfig = require('./webpack.base.config.js')

const path = require('path')
const resolve = path.resolve

const webpack = require('webpack')
const merge = require('webpack-merge')
const MiniCssExtractPlugin = require('mini-css-extract-plugin')
const HtmlWebPackPlugin = require('html-webpack-plugin')
const { CleanWebpackPlugin } = require('clean-webpack-plugin')
const CopyWebpackPlugin = require('copy-webpack-plugin')
const UglifyJsPlugin = require('uglifyjs-webpack-plugin')
const OptimizeCssAssetsPlugin = require('optimize-css-assets-webpack-plugin')
const ScriptExtHtmlWebpackPlugin = require('script-ext-html-webpack-plugin')
const BundleAnalyzerPlugin = require('webpack-bundle-analyzer').BundleAnalyzerPlugin

let publicPath = `https://static.ws.126.net/163/f2e/${pkg.channel}/${pkg.name}/`

module.exports = merge(baseWebpackConfig, {
  mode: 'production',
  output: {
    path: resolve(__dirname, '../dist'),
    filename: 'static/js/[name].[contenthash:8].js',
    publicPath
  },
  module: {
    rules: [
      {
        test: /\.html$/,
        use: [{
          loader: 'html-loader',
          options: {
            minimize: false,
            attrs: [':src']
          }
        }]
      }
    ]
  },
  plugins: [
    new CleanWebpackPlugin(),
    // new BundleAnalyzerPlugin(), //打包分析工具，按需打开
    new MiniCssExtractPlugin({
      filename: "static/css/[name].[contenthash:8].css"
    }),
    new HtmlWebPackPlugin({
        template: path.resolve(__dirname, "../src/index.html"),
        filename: "index.html"
    }),
    new ScriptExtHtmlWebpackPlugin({
      inline: /manifest\..*\.js$/,
      async: /main\..*\.js$/,
    }),
    new webpack.NamedModulesPlugin(),
    new webpack.DefinePlugin({
        'process.env.BASE_URL': JSON.stringify(publicPath),
        'process.env.ANT_PROJECT_ID': JSON.stringify(pkg.projectId),
        'process.env.PROJECT_CHANNEL': JSON.stringify(pkg.channel),
        'process.env.PROJECT_NAME': JSON.stringify(pkg.name)
    }),
    new CopyWebpackPlugin([
      {
        from: path.resolve(__dirname, '../static'),
        to: path.resolve(__dirname, '../dist/static'),
        ignore: ['.*']
      }
    ]),
  ],
  optimization: {
    minimizer:[
      new UglifyJsPlugin({
        exclude: /\.min\.js$/,
        cache: true,
        parallel: true,
        sourceMap: false,
        extractComments: false,
        uglifyOptions: {
          compress: {
            unused: true,
            drop_debugger: true
          },
          output: {
            comments: false
          }
        }
      }),
      new OptimizeCssAssetsPlugin({
        assetNameRegExp: /\.css$/g,
        cssProcessorOptions: {
          safe: true,
          autoprefixer: { disable: true },
          mergeLonghand: false,
          discardComments: {
            removeAll: true
          }
        },
        canPrint: true
      })
    ],
    splitChunks: {
      cacheGroups: {
        vendors: {
          test: /node_modules/,
          priority: 10,
          chunks: 'initial',
          name: 'vendors'
        },
        common: {
          name: 'common'
        }
      }
    },
    runtimeChunk: {
      name: 'manifest'
    }
  }
})

script/webpack.dev.config.js
/*
 * @Author: wanghaiyan
 * @Date: 2022-02-08 11:22:50
 * @LastEditTime: 2022-02-08 14:00:20
 * @LastEditors: wanghaiyan
 */
const pkg = require('../package.json')
const baseWebpackConfig = require('./webpack.base.config.js')

const path = require('path')
const resolve = path.resolve

const webpack = require('webpack')
const merge = require('webpack-merge')
const MiniCssExtractPlugin = require('mini-css-extract-plugin')
const HtmlWebPackPlugin = require('html-webpack-plugin')

const dev = process.env.NODE_ENV !== 'production'

module.exports = merge(baseWebpackConfig, {
  mode: 'development',
  output: {
    path: resolve(__dirname, 'dist'),
    filename: 'static/js/[name].js'
  },
  devtool: 'inline-source-map',
  devServer: {
    disableHostCheck: true,
    stats: {
      assets: false,
      colors: true,
      version: false,
      hash: false,
      timings: false,
      chunks: false,
      chunkModules: false
    },
    proxy: {
      // 代理
      // '/api': {
      //   target: 'https://test.3g.163.com/ug'
      // }
    }
  },
  module: {
    rules: [
      {
        test: /\.html$/,
        use: [{
          loader: 'html-loader',
          options: {
            minimize: false,
            attrs: [':src']
          }
        },{
          loader: 'ne-ssi-loader',
          options: {
            remote: {
              locations: 'https://news.163.com',
              charset: 'utf-8'
            }
          }
        },{
          loader: 'ne-velocity-loader'
        }]
      },
    ]
  },
  plugins: [
    new MiniCssExtractPlugin({
      filename: './static/css/[name].css'
    }),
    new HtmlWebPackPlugin({
      template: './src/index.html',
      filename: './index.html'
    }),
    new webpack.NamedModulesPlugin(),
    new webpack.DefinePlugin({
      'process.env.BASE_URL': JSON.stringify('/'),
      'process.env.ANT_PROJECT_ID': JSON.stringify(pkg.projectId)
    })
  ]
})
